### DISCUSSION TITLE
raylib v5.5

### DISCUSSION THREAD

**Discussion by raysan5** [raysan5/raylib#4503] [General]
![raylib55_banner](https://github.com/user-attachments/assets/2d18c1d8-e623-4f55-8496-bf6adfdb73df)

One year after raylib 5.0 release, arribes `raylib 5.5`, the next big revision of the library. It's been **11 years** since raylib 1.0 release and in all this time it has never stopped growing and improving. With an outstanding number of new contributors and improvements, it's, again, the biggest raylib release to date.

Some numbers for this release:

 - **+270** closed issues (for a TOTAL of **+1810**!)
 - **+800** commits since previous RELEASE (for a TOTAL of **+7770**!)
 - **+30** functions ADDED to raylib API (for a TOTAL of **580**!)
 - **+110** functions REVIEWED with fixes and improvements
 - **+140** new contributors (for a TOTAL of **+640**!)

Highlights for `raylib 5.5`:

 - **`NEW` raylib pre-configured Windows package**: The new raylib **portable and self-contained Windows package** for `raylib 5.5`, intended for nobel devs that start in programming world, comes with one big addition: support for **C code building for Web platform with one-single-mouse-click!** For the last 10 years, the pre-configured raylib Windows package allowed to edit simple C projects on Notepad++ and easely compile Windows executables with an automatic script; this new release adds the possibility to compile the same C projects for Web platform with a simple mouse click. This new addition **greatly simplifies C to WebAssembly project building for new users**. The `raylib Windows Installer` package can be downloaded for free from [raylib on itch.io](https://raysan5.itch.io/raylib).

 - **`NEW` raylib project creator tool**: A brand new tool developed to help raylib users to **setup new projects in a professional way**. `raylib project creator` generates a complete project structure with **multiple build systems ready-to-use** and **GitHub CI/CD actions pre-configured**. It only requires providing some C files and basic project parameters! The tools is [free and open-source](https://raysan5.itch.io/raylib-project-creator), and [it can be used online](https://raysan5.itch.io/raylib-project-creator)!.

 - **`NEW` Platform backend supported: RGFW**: Thanks to the `rcore` platform-split implemented in `raylib 5.0`, **adding new platforms backends has been greatly simplified**, new backends can be added using provided template, self-contained in a single C module, completely portable. A new platform backend has been added: [`RGFW`](https://github.com/raysan5/raylib/blob/master/src/platforms/rcore_desktop_rgfw.c). `RGFW` is a **new single-file header-only portable library** ([`RGFW.h`](https://github.com/ColleagueRiley/RGFW)) intended for platform-functionality management (windowing and inputs); in this case for **desktop platforms** (Windows, Linux, macOS) but also for **Web platform**. It adds a new alternative to the already existing `GLFW` and `SDL` platform backends.
 
 - **`NEW` Platform backend version supported: SDL3**: Previous `raylib 5.0` added support for `SDL2` library, and `raylib 5.5` not only improves SDL2 functionality, with several issues reviewed, but also adds support for the recently released big SDL update in years: [`SDL3`](https://wiki.libsdl.org/SDL3/FrontPage). Now users can **select at compile time the desired SDL version to use**, increasing the number of potential platforms supported in the future!
 
 - **`NEW` Retro-console platforms supported: Dreamcast, N64, PSP, PSVita, PS4**: Thanks to the platform-split on `raylib 5.0`, **supporting new platform backends is easier than ever!** Along the raylib `rlgl` module support for the  `OpenGL 1.1` graphics API, it opened the door to [**multiple homebrew retro-consoles backend implementations!**](https://github.com/raylib4Consoles) It's amazing to see raylib running on +20 year old consoles like [Dreamcast](https://github.com/raylib4Consoles/raylib4Dreamcast), [PSP](https://github.com/raylib4Consoles/raylib4Psp) or [PSVita](https://github.com/psp2dev/raylib4Vita), considering the hardware constraints of those platforms and proves **raylib outstanding versability!** Those additional platforms can be found in separate repositories and have been created by the amazing programmer Antonio Jose Ramos Marquez (@psxdev).
 
 - **`NEW` GPU Skinning support**: After lots of requests for this feature, it has been finally added to raylib thanks to the contributor Daniel Holden (@orangeduck), probably the developer that has further pushed models animations with raylib, developing two amazing tools to visualize and test animations: [GenoView](https://github.com/orangeduck/GenoView) and [BVHView](https://github.com/orangeduck/BVHView). Adding GPU skinning was a tricky feature, considering it had to be **available for all raylib supported platforms**, including limited ones like Raspberry Pi with OpenGL ES 2.0, where some advance OpenGL features are not available (UBO, SSBO, Transform Feedback) but a multi-platform solution was found to make it possible. A new example, [`models_gpu_skinning`](https://github.com/raysan5/raylib/blob/master/examples/models/models_gpu_skinning.c) has been added to illustrate this new functionality. As an extra, previous existing CPU animation system has been greatly improved, multiplying performance by a factor (simplifiying required maths).
 
 - **`NEW` [`raymath`](https://github.com/raysan5/raylib/blob/master/src/raymath.h) C++ operators**: After several requested for this feature, C++ math operators for `Vector2`, `Vector3`, `Vector4`, `Quaternion` and `Matrix` has been added to `raymath` as an extension to current implementation. Despite being only available for C++ because C does not support it, these operators **simplify C++ code when doing math operations**.

Beside those new big features, `raylib 5.5` comes with MANY other improvements: 

- Normals support on batching system
- Clipboard images reading support
- CRC32/MD5/SHA1 hash computation
- Gamepad vibration support
- Improved font loading (no GPU required) with BDF fonts support
- Time-based camera movement
- Improved GLTF animations loading

...and [much much more](https://github.com/raysan5/raylib/blob/master/CHANGELOG), including **many functions reviews and new functions added!**
 
Make sure to check raylib [CHANGELOG](https://github.com/raysan5/raylib/blob/master/CHANGELOG) for a detailed list of changes!

To end with, I want to **thank all the contributors (+640!**) that along the years have **greatly improved raylib** and pushed it further and better day after day. Thanks to all of them, raylib is the amazing library it is today.

Last but not least, I want to thank **raylib sponsors and all the raylib community** for their support and continuous engagement with the library, creating and sharing amazing raylib projects on a daily basis. **Thanks for making raylib a great platform to enjoy games/tools/graphic programming!**

**After 11 years of development, `raylib 5.5` is the best raylib ever.**

**Enjoy programming with raylib!** :)

## New Contributors
* @Kimo-s made their first contribution in https://github.com/raysan5/raylib/pull/3528
* @MrScautHD made their first contribution in https://github.com/raysan5/raylib/pull/3546
* @AuzFox made their first contribution in https://github.com/raysan5/raylib/pull/3552
* @Zapunidi made their first contribution in https://github.com/raysan5/raylib/pull/3555
* @ManuMario0 made their first contribution in https://github.com/raysan5/raylib/pull/3574
* @Minmoose made their first contribution in https://github.com/raysan5/raylib/pull/3609
* @davidthings made their first contribution in https://github.com/raysan5/raylib/pull/3615
* @riadbettole made their first contribution in https://github.com/raysan5/raylib/pull/3624
* @lipx1508 made their first contribution in https://github.com/raysan5/raylib/pull/3618
* @benjibst made their first contribution in https://github.com/raysan5/raylib/pull/3627
* @rmn20 made their first contribution in https://github.com/raysan5/raylib/pull/3631
* @orosmatthew made their first contribution in https://github.com/raysan5/raylib/pull/3632
* @Starpelly made their first contribution in https://github.com/raysan5/raylib/pull/3640
* @devdad made their first contribution in https://github.com/raysan5/raylib/pull/3621
* @DongkunLee made their first contribution in https://github.com/raysan5/raylib/pull/3536
* @Toctave made their first contribution in https://github.com/raysan5/raylib/pull/3585
* @cinghycreations made their first contribution in https://github.com/raysan5/raylib/pull/3641
* @tromero made their first contribution in https://github.com/raysan5/raylib/pull/3666
* @wisonye made their first contribution in https://github.com/raysan5/raylib/pull/3682
* @maverikou made their first contribution in https://github.com/raysan5/raylib/pull/3678
* @IoIxD made their first contribution in https://github.com/raysan5/raylib/pull/3681
* @seiren-games made their first contribution in https://github.com/raysan5/raylib/pull/3692
* @casavaca made their first contribution in https://github.com/raysan5/raylib/pull/3710
* @Minnowo made their first contribution in https://github.com/raysan5/raylib/pull/3712
* @LievenPetersen made their first contribution in https://github.com/raysan5/raylib/pull/3720
* @candrewlee14 made their first contribution in https://github.com/raysan5/raylib/pull/3727
* @karl-zylinski made their first contribution in https://github.com/raysan5/raylib/pull/3701
* @Blockguy24 made their first contribution in https://github.com/raysan5/raylib/pull/3732
* @JayLCypher made their first contribution in https://github.com/raysan5/raylib/pull/3737
* @idircarlos made their first contribution in https://github.com/raysan5/raylib/pull/3751
* @mllimo made their first contribution in https://github.com/raysan5/raylib/pull/3750
* @aubs-dev made their first contribution in https://github.com/raysan5/raylib/pull/3760
* @marrony made their first contribution in https://github.com/raysan5/raylib/pull/3770
* @bmbkr made their first contribution in https://github.com/raysan5/raylib/pull/3774
* @stanthesoupking made their first contribution in https://github.com/raysan5/raylib/pull/3735
* @The-Night-Watch made their first contribution in https://github.com/raysan5/raylib/pull/3776
* @oblerion made their first contribution in https://github.com/raysan5/raylib/pull/3771
* @jfoscarini made their first contribution in https://github.com/raysan5/raylib/pull/3799
* @BliznyukNM made their first contribution in https://github.com/raysan5/raylib/pull/3804
* @Logicless-Coder made their first contribution in https://github.com/raysan5/raylib/pull/3817
* @GideonSerf made their first contribution in https://github.com/raysan5/raylib/pull/3819
* @Dev-Tade made their first contribution in https://github.com/raysan5/raylib/pull/3832
* @GaryMcWhorter made their first contribution in https://github.com/raysan5/raylib/pull/3823
* @ipzaur made their first contribution in https://github.com/raysan5/raylib/pull/3833
* @Bowserinator made their first contribution in https://github.com/raysan5/raylib/pull/3828
* @hardliner66 made their first contribution in https://github.com/raysan5/raylib/pull/3835
* @zuckschwerdt made their first contribution in https://github.com/raysan5/raylib/pull/3839
* @aiafrasinei made their first contribution in https://github.com/raysan5/raylib/pull/3842
* @REDl3east made their first contribution in https://github.com/raysan5/raylib/pull/3846
* @joyousblunder made their first contribution in https://github.com/raysan5/raylib/pull/3868
* @zyperpl made their first contribution in https://github.com/raysan5/raylib/pull/3871
* @proberge-dev made their first contribution in https://github.com/raysan5/raylib/pull/3881
* @MrMugame made their first contribution in https://github.com/raysan5/raylib/pull/3879
* @ProIcons made their first contribution in https://github.com/raysan5/raylib/pull/3891
* @iarkn made their first contribution in https://github.com/raysan5/raylib/pull/3896
* @Belllg made their first contribution in https://github.com/raysan5/raylib/pull/3901
* @AriaMoKr made their first contribution in https://github.com/raysan5/raylib/pull/3910
* @He-Is-HaZaRdOuS made their first contribution in https://github.com/raysan5/raylib/pull/3908
* @Mute124 made their first contribution in https://github.com/raysan5/raylib/pull/3914
* @dylanlangston made their first contribution in https://github.com/raysan5/raylib/pull/3915
* @benjitrosch made their first contribution in https://github.com/raysan5/raylib/pull/3919
* @eldskald made their first contribution in https://github.com/raysan5/raylib/pull/3923
* @dertseha made their first contribution in https://github.com/raysan5/raylib/pull/3907
* @KotzaBoss made their first contribution in https://github.com/raysan5/raylib/pull/3912
* @lima-limon-inc made their first contribution in https://github.com/raysan5/raylib/pull/3938
* @OetkenPurveyorOfCode made their first contribution in https://github.com/raysan5/raylib/pull/3939
* @UmgefallenesGlas made their first contribution in https://github.com/raysan5/raylib/pull/3949
* @gabriel-marques made their first contribution in https://github.com/raysan5/raylib/pull/3956
* @ColleagueRiley made their first contribution in https://github.com/raysan5/raylib/pull/3941
* @alexmozaidze made their first contribution in https://github.com/raysan5/raylib/pull/3979
* @Filyus made their first contribution in https://github.com/raysan5/raylib/pull/3974
* @myQwil made their first contribution in https://github.com/raysan5/raylib/pull/3977
* @CosmicBagel made their first contribution in https://github.com/raysan5/raylib/pull/3983
* @cemalgnlts made their first contribution in https://github.com/raysan5/raylib/pull/3940
* @FishingHacks made their first contribution in https://github.com/raysan5/raylib/pull/3986
* @ListeriaM made their first contribution in https://github.com/raysan5/raylib/pull/3994
* @Odex64 made their first contribution in https://github.com/raysan5/raylib/pull/3995
* @sgalindo made their first contribution in https://github.com/raysan5/raylib/pull/3990
* @L-Briand made their first contribution in https://github.com/raysan5/raylib/pull/4004
* @avx0 made their first contribution in https://github.com/raysan5/raylib/pull/4011
* @DarkAssassin23 made their first contribution in https://github.com/raysan5/raylib/pull/4013
* @vaezim made their first contribution in https://github.com/raysan5/raylib/pull/4017
* @kai-z99 made their first contribution in https://github.com/raysan5/raylib/pull/4018
* @denovodavid made their first contribution in https://github.com/raysan5/raylib/pull/4022
* @paulmelis made their first contribution in https://github.com/raysan5/raylib/pull/4034
* @Sprixitite made their first contribution in https://github.com/raysan5/raylib/pull/4020
* @jgabaut made their first contribution in https://github.com/raysan5/raylib/pull/4039
* @KonradGrande made their first contribution in https://github.com/raysan5/raylib/pull/4040
* @fruzitent made their first contribution in https://github.com/raysan5/raylib/pull/4042
* @carverdamien made their first contribution in https://github.com/raysan5/raylib/pull/4049
* @lzralbu made their first contribution in https://github.com/raysan5/raylib/pull/4054
* @VitoTringolo made their first contribution in https://github.com/raysan5/raylib/pull/4053
* @TokyoSU made their first contribution in https://github.com/raysan5/raylib/pull/4059
* @ShalokShalom made their first contribution in https://github.com/raysan5/raylib/pull/4068
* @okvik made their first contribution in https://github.com/raysan5/raylib/pull/4027
* @spelufo made their first contribution in https://github.com/raysan5/raylib/pull/3744
* @CrackedPixel made their first contribution in https://github.com/raysan5/raylib/pull/4074
* @jspast made their first contribution in https://github.com/raysan5/raylib/pull/4104
* @brccabral made their first contribution in https://github.com/raysan5/raylib/pull/4105
* @FrankyBalu made their first contribution in https://github.com/raysan5/raylib/pull/4129
* @ninadsachania made their first contribution in https://github.com/raysan5/raylib/pull/4136
* @NishiOwO made their first contribution in https://github.com/raysan5/raylib/pull/4139
* @InventorXtreme made their first contribution in https://github.com/raysan5/raylib/pull/4150
* @kai-kj made their first contribution in https://github.com/raysan5/raylib/pull/4154
* @jkaup made their first contribution in https://github.com/raysan5/raylib/pull/4102
* @CDM15y made their first contribution in https://github.com/raysan5/raylib/pull/4167
* @jaens made their first contribution in https://github.com/raysan5/raylib/pull/4175
* @Jutastre made their first contribution in https://github.com/raysan5/raylib/pull/4179
* @SoloByte made their first contribution in https://github.com/raysan5/raylib/pull/4189
* @maxmutant made their first contribution in https://github.com/raysan5/raylib/pull/4184
* @rnpnr made their first contribution in https://github.com/raysan5/raylib/pull/4197
* @ZzzhHe made their first contribution in https://github.com/raysan5/raylib/pull/4186
* @lnc3l0t made their first contribution in https://github.com/raysan5/raylib/pull/4193
* @MaximKn1 made their first contribution in https://github.com/raysan5/raylib/pull/4202
* @hanaxars made their first contribution in https://github.com/raysan5/raylib/pull/4247
* @Paperdomo101 made their first contribution in https://github.com/raysan5/raylib/pull/4270
* @Bugsia made their first contribution in https://github.com/raysan5/raylib/pull/4276
* @Mercotui made their first contribution in https://github.com/raysan5/raylib/pull/4275
* @konstruktor227 made their first contribution in https://github.com/raysan5/raylib/pull/4279
* @satchelfrost made their first contribution in https://github.com/raysan5/raylib/pull/4203
* @Tchan0 made their first contribution in https://github.com/raysan5/raylib/pull/4284
* @IllusionMan1212 made their first contribution in https://github.com/raysan5/raylib/pull/4288
* @emdzej made their first contribution in https://github.com/raysan5/raylib/pull/4305
* @masnm made their first contribution in https://github.com/raysan5/raylib/pull/4308
* @Alex-Velez made their first contribution in https://github.com/raysan5/raylib/pull/4311
* @dostoievsky made their first contribution in https://github.com/raysan5/raylib/pull/4318
* @SusgUY446 made their first contribution in https://github.com/raysan5/raylib/pull/4310
* @foxblock made their first contribution in https://github.com/raysan5/raylib/pull/4302
* @orangeduck made their first contribution in https://github.com/raysan5/raylib/pull/4321
* @asdqwe made their first contribution in https://github.com/raysan5/raylib/pull/4334
* @Ridge3Dproductions made their first contribution in https://github.com/raysan5/raylib/pull/4332
* @FluxFlu made their first contribution in https://github.com/raysan5/raylib/pull/4356
* @VisenDev made their first contribution in https://github.com/raysan5/raylib/pull/4366
* @HarryDC made their first contribution in https://github.com/raysan5/raylib/pull/4373
* @cedeon made their first contribution in https://github.com/raysan5/raylib/pull/4374
* @yuval-herman made their first contribution in https://github.com/raysan5/raylib/pull/4383
* @R-YaTian made their first contribution in https://github.com/raysan5/raylib/pull/4384
* @Joshalosh made their first contribution in https://github.com/raysan5/raylib/pull/4382
* @arrecis made their first contribution in https://github.com/raysan5/raylib/pull/4391
* @rapha-s made their first contribution in https://github.com/raysan5/raylib/pull/4408
* @cypressru made their first contribution in https://github.com/raysan5/raylib/pull/4423
* @franztt made their first contribution in https://github.com/raysan5/raylib/pull/4421
* @juliohq made their first contribution in https://github.com/raysan5/raylib/pull/4428
* @mpv-enjoyer made their first contribution in https://github.com/raysan5/raylib/pull/4448
* @evertonse made their first contribution in https://github.com/raysan5/raylib/pull/4435
* @archewashi made their first contribution in https://github.com/raysan5/raylib/pull/4451
* @MikiZX1 made their first contribution in https://github.com/raysan5/raylib/pull/4460
* @waveydave made their first contribution in https://github.com/raysan5/raylib/pull/4461
* @decromo made their first contribution in https://github.com/raysan5/raylib/pull/4462
* @deathbeam made their first contribution in https://github.com/raysan5/raylib/pull/4468
* @kimierik made their first contribution in https://github.com/raysan5/raylib/pull/4485
* @OussamaTeyib made their first contribution in https://github.com/raysan5/raylib/pull/4493
* @zet23t made their first contribution in https://github.com/raysan5/raylib/pull/4494

**Full Changelog**: https://github.com/raysan5/raylib/compare/5.0...5.5

<hr /><em>This discussion was created from the release <a href='https://github.com/raysan5/raylib/releases/tag/5.5'>raylib v5.5</a>.</em>

**Comment by GyanPrakash2483** [comment]
I would like to thank and congratulate @raysan5 and all contributors of raylib for this release, thakyou you all to develop and maintain my favourate game development library.

**Reply by sajibsrs** [reply]
Me too, it significantly reduced the CPU usage, at least for me in the Mac Apple silicon m2. From 8-10% to 4.8-5.4%. There's slight memory usage improvement too. Thanks a lot.

**Comment by girng** [comment]
incredible ♥

**Reply by JesseLade** [reply]
Raylib is an AMAZING c++ library i love it how was it made?

**Comment by Sparklight77** [comment]
Raylib is the one absolutely outstanding game dev Library out there. By the way I can't see armv7/armhf builds, maybe that comes next time!

**Reply by apiva** [reply]
Hi, getting raylib to build from source on armv7/armhf is easy peasy. I do that to get my games running on my Powkiddy (that really is a Raspberry Pi). You can even to that in Windows using WSL! I'll copy&paste my notes here.

## Install Debian WSL

Install Debian through WSL:
```
wsl --install Debian
```

Update the distro:
```
sudo apt update
```
```
sudo apt upgrade
```
## Compile for Linux

Install C++ development tools
```
sudo apt install git g++ cmake ninja-build
```

Install the libraries needed by Raylib
```
sudo apt install libasound2-dev libx11-dev libxrandr-dev libxi-dev libgl1-mesa-dev libglu1-mesa-dev libxcursor-dev libxinerama-dev
```

Go the the game folder and run CMake:
```
cmake -S . -B temp/build/linux -D CMAKE_BUILD_TYPE=Release
```

Build the game:
```
cmake --build temp/build/linux
```
## Compile for Raspberry Pi

Add the `armhf` architecture for packages:
```
sudo dpkg --add-architecture armhf
```
```
sudo apt update
```

Get the compiler:
```
sudo apt install g++-arm-linux-gnueabihf
```

Get the dependencies:
```
sudo apt install libdrm-dev:armhf libegl1-mesa-dev:armhf libgles2-mesa-dev:armhf libgbm-dev:armhf
```

Go the the game folder and run CMake:
```
cmake -S . -B temp/build/raspberrypi -D CMAKE_C_COMPILER=arm-linux-gnueabihf-gcc -D CMAKE_CXX_COMPILER=arm-linux-gnueabihf-g++ -D PLATFORM=DRM -D CMAKE_BUILD_TYPE=Release
```

Build the game:
```
cmake --build temp/build/raspberrypi
```

## Compile for Raspberry Pi 64 bits

Add the `arm64` architecture for packages:
```
sudo dpkg --add-architecture arm64
```
```
sudo apt update
```

Get the compiler:
```
sudo apt install g++-aarch64-linux-gnu
```

Get the dependencies:
```
sudo apt install libdrm-dev:arm64 libegl1-mesa-dev:arm64 libgles2-mesa-dev:arm64 libgbm-dev:arm64
```

Go the the game folder and run CMake:
```
cmake -S . -B temp/build/raspberrypi64 -D CMAKE_C_COMPILER=aarch64-linux-gnu-gcc -D CMAKE_CXX_COMPILER=sta -D PLATFORM=DRM -D CMAKE_BUILD_TYPE=Release
```

Build the game:
```
cmake --build temp/build/raspberrypi64
```

## References
[Raspberry Pi Forums: Compile raylib example for raspberry PI](https://forums.raspberrypi.com/viewtopic.php?t=250928)
[Raylib Wiki: Working on Raspberry Pi](https://github.com/raysan5/raylib/wiki/Working-on-Raspberry-Pi)
[# Cross compiling for arm or aarch64 on Debian or Ubuntu](https://jensd.be/1126/linux/cross-compiling-for-arm-or-aarch64-on-debian-or-ubuntu)

**Reply by Sparklight77** [reply]
Tysm but I already built it :)

**Comment by JesseLade** [comment]
Raylib is an AMAZING c++ library i love it how was it made?

**Reply by Sparklight77** [reply]
It's not cpp, it was written in C99

**Comment by JesseLade** [comment]
ITS AMAZING

**Comment by ElvintsY** [comment]
I'm a beginner trying to use raylib to make a small game, am got a question, why does StopSoundMulti doesn't exist in the latest version (5.5) of raylib for mingw64 compiler or is that function unnecessary? Just curious, I know I can just use StopSound for whatever I'm doing.

**Comment by paulxcalder** [comment]
is there any guide how to use it? i installed this libraary but no idea how to use
visual studio code and C language

**Reply by maiconpintoabreu** [reply]
https://github.com/raysan5/raylib/wiki/Using-raylib-in-VSCode
